File: code_samples/haskell/file.hs
{-# LANGUAGE DeriveFunctor, DeriveFoldable, DeriveTraversable #-}

import Data.Map (Map, fromList, toList)
import Data.List (foldl')
import Text.Printf (printf)

-- Define a data type to hold complex data
data ComplexData = ComplexData
  { id :: Int
  , name :: String
  , values :: [Int]
  } deriving (Show)

-- Create a new instance of ComplexData
newComplexData :: Int -> String -> ComplexData
newComplexData id name = ComplexData id name [id * 10, id * 20, id * 30]

-- Calculate the average of values in ComplexData
calculate :: ComplexData -> Double
calculate (ComplexData _ _ values) = 
  fromIntegral (sum values) / fromIntegral (length values)

-- Generate a map of ComplexData instances
generateData :: Int -> Map Int ComplexData
generateData count = fromList [(id, newComplexData id ("Item " ++ show id)) | id <- [1..count]]

-- Save data to a file
saveToFile :: FilePath -> Map Int ComplexData -> IO ()
saveToFile filename data = 
  writeFile filename (unlines (map formatData (toList data)))

-- Format data as a string
formatData :: (Int, ComplexData) -> String
formatData (_, ComplexData id name values) =
  printf "ID: %d, Name: %s, Average Value: %.2f" id name (calculate (ComplexData id name values))

-- Main function to simulate the processing and saving of data
main :: IO ()
main = do
  putStrLn "Processing data..."
  let data = generateData 10
  putStrLn "\nGenerated Data:"
  mapM_ (putStrLn . formatData) (toList data)

  saveToFile "data_output.txt" data
  putStrLn "\nData successfully saved to 'data_output.txt'."

  -- Simulate async-like behavior with a delay
  putStrLn "\nSimulating async-like behavior..."
  let delay = 1000 * 1000 -- 1 second in microseconds
  threadDelay delay
  putStrLn "Task completed!"

File: code_samples/elixir/file.ex
defmodule DataProcessor do
  defmodule ComplexData do
    defstruct [:id, :name, :values]

    def new(id, name) do
      values = Enum.map([id * 10, id * 20, id * 30], &(&1))
      %ComplexData{id: id, name: name, values: values}
    end

    def calculate(%ComplexData{values: values}) do
      Enum.sum(values) / length(values)
    end
  end

  def generate_data(count) do
    1..count
    |> Enum.map(fn id -> {id, ComplexData.new(id, "Item #{id}")} end)
    |> Enum.into(%{})
  end

  def save_to_file(filename, data) do
    File.write!(filename, Enum.map(data, fn {_, item} ->
      "ID: #{item.id}, Name: #{item.name}, Average Value: #{ComplexData.calculate(item)}"
    end) |> Enum.join("\n"))
  end

  def process_data(count) do
    IO.puts("Processing data...")
    Process.sleep(2000)

    data = generate_data(count)

    IO.puts("\nGenerated Data:")
    Enum.each(data, fn {_, item} ->
      IO.puts("#{item.id}: #{item.name} => Average: #{ComplexData.calculate(item)}")
    end)

    save_to_file("data_output.txt", data)
    IO.puts("\nData successfully saved to 'data_output.txt'.")

    IO.puts("\nSimulating async-like behavior...")
    Process.sleep(1000)
    IO.puts("Task completed!")
  end
end

DataProcessor.process_data(10)
